### ISSUE TITLE
Support CE customised built-in inheritance

### ISSUE THREAD

**Issue by jmalonzo** [WebReflection/custom-tag#1]
Hi! Are there any plans to support CE inheritance?

**Comment by WebReflection** [comment]
Just use extends passing the class you'd like inherit should work, doesn't it?

**Comment by jmalonzo** [comment]
Sorry, I meant to say inheriting a customised built-in CE. For example

```
class A extends HTMLTableElement {}

class B extends A {}
```

At the moment, it checks whether 'extends' is a string or not to determine whether to pass {extends: 'foo'} to customElements.define.

Though, I'm not sure if the example above is even allowed in the spec?

Thanks.

**Comment by WebReflection** [comment]
It's written in the README , extends accepts string OR class . You can extend 'table' or HTMLTableElement.

**Comment by WebReflection** [comment]
You can extend A too, but these have to be Custom Elements and these need a name.

Extending builtins is OK here but you need my document-register-element polyfill upfront

**Comment by jmalonzo** [comment]
No worries, I'll try to figure it out.

Thanks a lot!

**Comment by WebReflection** [comment]
Apologies, quick replies from the phone and surely not so good documentation

**Comment by jmalonzo** [comment]
@WebReflection That's fine. I've created a jsbin to make it easier - http://jsbin.com/hahuvitiho/edit?html,console,output

As you see TableB doesn't render, but changing `extends: TableA` to `extends: table` works as expected. I'm just not sure if this is a custom-tag issue or a document-register-element issue, or a conceptual issue with what I'm trying to do.

Thanks

**Comment by WebReflection** [comment]
Ha, interesting example.

First of all, a workaround that works:
```js
    let TableA = CustomTag({
      name: 'table-a',
      extends: 'table',
      onConnect() {
        console.log('table-a connect');
      },
      show() {
        console.log('table-a show');
      }
    });

    // use Classtrophobic directly
    const TableB = Class({
      extends: TableA,
      onConnect() {
        this.super.onConnect();
        console.log('table-b connect');
        let tr = document.createElement('tr');
        let td = document.createElement('td');
        td.innerText = '1';
        tr.appendChild(td);
        this.appendChild(tr);
      }
    });

    customElements.define('table-b', TableB, {extends: 'table'});
```

**or** use directly ES2015
```js
    class TableB extends TableA {
      onConnect() {
        super.onConnect();
        console.log('table-b connect');
        let tr = document.createElement('tr');
        let td = document.createElement('td');
        td.innerText = '1';
        tr.appendChild(td);
        this.appendChild(tr);
      }
    }

    customElements.define('table-b', TableB, {extends: 'table'});
```

The potential fix is not as trivial as it looks. Following the list of issues:

  * if you extend `table-a` as string, that would work as native extend only if you then have this kind of layout: `<table-b is="table-a">`
  * the current CE specification gives a way to retrieve a class via string, but IIRC not vice-versa. This means that if you pass `TableA` constructor I cannot know if that was extending a native or not. The registry lacks some extra method for better introspection
  * the kind of inheritance you want is by using `table` upgrading it up to N sub-classes still keeping the original tag-name ... I guess this is one of the quirks vendors were talking about the `is=""` attribute.

I've never seen the need to subclass an already subclassed native but I see it as a valid use case.

**TL;DR** I am reopening this as actual bug and I'll try to figure out if it's possible to implement this pattern on top, without bloating and without causing other side effects.

However, consider CustomTag was born to mainly fix at (at least) one level most common CE pitfalls.

I don't see the usage of Classtrophobic or even native classes on top of a CustomTag base as a real blocker, actually I believe it's a better pattern for future refactoring so there's only one level to change on the root of the inheritance, instead of N classes.

Thoughts ? Thanks for providing a valid example.

**Comment by jmalonzo** [comment]
Agree with the list of issues, and the fix being non-trivial.

I tried the workaround and I had more success with using native classes, as classtrophobic seems to throw an exception - http://jsbin.com/wukihofoci/1/edit?html,output

The following code throws an exception:
```
   const TableC = Class({
      extends: TableA,
      connectedCallback() {
        this.super.onConnect();
        console.log('table-c connect');
        let tr = document.createElement('tr');
        let td = document.createElement('td');
        td.innerText = '1';
        tr.appendChild(td);
        this.appendChild(tr);
        this.super.show();
      }
    });
    customElements.define('table-c', TableC, {extends: 'table'});
```

It seems to throw an error around this line - https://github.com/WebReflection/classtrophobic/blob/master/classtrophobic.js#L44

I actually don't mind if CustomTag stays as it is as long as I can use Classtrophobic (I can't use native classes as I'll encounter transpiling issues with Babel which Classtrophobic is trying to fix).

**Comment by WebReflection** [comment]
I'm on a train without tethering options. Can you please paste the non working code in here since jsbin on safari mobile is quite useless? Thanks

**Comment by jmalonzo** [comment]
Sorry, I had to update the comment above as I had to fix the link to the jsbin. I've included the non-working code above as well.

**Comment by WebReflection** [comment]
Use onConnect instead, you are extending a CustomTag class ;-)

**Comment by WebReflection** [comment]
Sorry for the unconventional assistance..,

To recap, this is using CustomTag as base class:
![img_1809](https://cloud.githubusercontent.com/assets/85749/26629872/97ab1f4c-4604-11e7-9fe6-b1b7ba4856c1.JPG)

While this is avoiding CustomTag as a whole and using Classtrophobic only
![img_1808](https://cloud.githubusercontent.com/assets/85749/26629910/bb693022-4604-11e7-905f-7b176fbec222.JPG)

I hope you can read those examples 😅

**Comment by jmalonzo** [comment]
Thanks so much for your help. Yep, those seems to work fine :-)

**Comment by WebReflection** [comment]
To better understand why your example wasn't working, if you call a super method that has never been defined, and indeed CustomTag doesn't use native callbacks, a super.method call would fail like in any native class example.

**Comment by jmalonzo** [comment]
Yeah, that makes sense now. Thank you!
